### ISSUE TITLE
Normalize Space dimensions to (0, 1) in Optimizer

### ISSUE THREAD

**Issue by MechCoder** [scikit-optimize/scikit-optimize#400]
It seems like we do not normalize the dimensions of the space in the ask-and-tell API. Noted by @cranmer in https://github.com/scikit-optimize/scikit-optimize/issues/399#issuecomment-307887516

**Comment by betatim** [comment]
If the `estimator` is a GP then we need to always normalise to (0, 1)? Asking because using `Optimizer` directly probably means you are more of an expert who wants to do exactly what they want with as little interference from the library. At least we should allow such a way of working. How do we make both easy switching from `*_minimize` to `Optimizer` possible and let experts do expert things?

**Comment by iaroslav-ai** [comment]
Just pitching in my 2 cents. 
Maybe using something like a pipeline as a surrogate could be a solution, for example:
```python
from skopt.learning import GaussianProcessRegressor
from skopt.space import Real
from skopt import Optimizer

import numpy as np

from sklearn.pipeline import Pipeline
from sklearn.preprocessing import MinMaxScaler

# modified Pipeline class to handle `return_std` etc argument for predict
class SurrogatePipeline(Pipeline):
    def predict(self, X, *args, **kwargs):
        Xt = X
        for name, transform in self.steps[:-1]:
            if transform is not None:
                Xt = transform.transform(Xt)
        return self.steps[-1][-1].predict(Xt, *args, **kwargs)

# optimizer that uses pipeline as a surrogate
opt = Optimizer(
    dimensions=[Real(-2.0, 2.0) for i in range(10)],
    base_estimator=SurrogatePipeline([
        ('Scaler', MinMaxScaler()),
        ('Surrogate', GaussianProcessRegressor()),
    ])
)

# example optimization
obj = lambda x: np.sum(x)

for i in range(32):
    p = opt.ask()
    opt.tell(p, obj(p))
    print(min(opt.yi))
```
Instead of `MinMaxScaler`, a dedicated scaler can be used that normalizes based on dimensions of a problem.  This should make it possible for experts to modify the pipeline if necessary to remove normalization. Maybe also nice in general in someone wants to have a complex surrogate pipeline.

**Comment by MechCoder** [comment]
> How do we make both easy switching from *_minimize to Optimizer possible and let experts do expert things?

Good point. For starters, we can maybe support `base_estimator` to be strings and for those cases we can handle things similar to what we do in our minimize functions.

**Comment by MechCoder** [comment]
I like the idea by @iaroslav-ai of moving the idea of transformations from the `Dimension` instance to the `base_estimator`. What are the cons?

**Comment by iaroslav-ai** [comment]
In my veiw such approach might result in a bit of a boilerplate code, and one would need to define an instance of array of dimensions, and then feed it to both Optimizer and surrogate pipeline (at least how I imagine it). So if someone new tries to use Optimizer, they might not figure out to use pipelines (but examples might be helpful for that). 

Maybe somewhat a boilerplate is a price for larger flexibility with Optimizer. 

On somewhat related note: with pipeline approach one can also try different normalization techniques, such as StandardScaler for example, maybe it turns out that it performs better than just normalizaiton to [0, 1]. Maybe benchmarking different scalers could reveal something interesting.

**Comment by iaroslav-ai** [comment]
Maybe something like `skopt.recipes` could reduce the boilerplate with pipeline approach, where you have a function that instantiates a pipeline with certain surrogate and normalizaiton. 

Again somewhat related idea: black box optimization can be applied to black box optimization in order to maximize for example average score on benchmarks by selecting type of normalization and surrogate (and maybe something else). This might result in some high performing "best" recipe. Seems like it might be something useful, fun and meta :)

**Comment by betatim** [comment]
I think the question is: is normalising a property of the search space or a property of the optimiser? Right now `skopt` thinks it is a property of the space. If you are using a GP based minimizer then you probably want to normalise all dimensions (except categorical) but you don't need that for a tree based one. So this would suggest really it is a property of the optimiser you are using?

**Comment by MechCoder** [comment]
I would suggest to keep things simple for now, i.e scale to (0, 1) irrespective of the minimizer and the advanced user do whatever he wants to do (by providing a Space instance)

**Comment by betatim** [comment]
If we can't think of a reason not to scale by default, should we change the default value of https://github.com/scikit-optimize/scikit-optimize/blob/master/skopt/space/space.py#L166 to take care of this (also for other `Dimension` classes)?

**Comment by MechCoder** [comment]
That is a good idea.

However for a long-term solution, I would suggest to decouple the idea of transformations from Dimension instances. A Dimension should just have a range and should be allowed to sample from that range.

We should have a pipeline that has a `Scaler` that does the necessary transformations before supplying it to the `base_estimator`. Not sure how easy it is to accomplish that.
